textx/lang.py

```diff
@@ -364,9 +366,13 @@ def _add_reffered_classes(rule, inh_by, start=False):
                     else:
                         _add_reffered_classes(rule, cls._tx_inh_by, start=True)
 
-        resolved_classes = set()
-        for cls in metamodel:
-            _determine_rule_type(cls)
+        # Multi-pass rule type resolving to support circular rule references.
+        has_change = [True]
```

Comment:
I believe the use of the array here is a trick to prevent accidental redefinitions of the variable, when assigning a new value. If this is not 100% standard, I suggest to put a small comment.

Reply:
Actually, this is a trick for Python 2 to support changing the value of outer scope variable. In Python 3 there is a keyword `nonlocal` which declares that the variable is to be found in the outer scope and the inner functions may actually set this reference. There is no such thing in Python 2. There is a kw `global` but it specifies that the variable is to be found in the global (module-level) scope. But the trick used here is that, although you can't set the variable from the outer scope, you can access/read them. Thus, the reference to the list `has_change` is read inside of the inner function and only the content of the list has been changed.

When we drop support for Python 2 one day, this can be changed to use `nonlocal`.

I've added a short comment in the code.